java/src/jmri/jmrix/openlcb/swing/stleditor/StlEditorPane.java

```diff
@@ -0,0 +1,2753 @@
+                            versionString,
+                            TOWER_LCC_Q_NODE_VERSION_STRING),
+                    Bundle.getMessage("TitleVersion"),
+                    JmriJOptionPane.WARNING_MESSAGE);
+            return false;
+        }
+
+        return true;
+    }
+
+    // Notifies that the contents of a given entry have changed. This will delete and re-add the
+    // entry to the model, forcing a refresh of the box.
+    public void updateComboBoxModelEntry(NodeEntry nodeEntry) {
+        int idx = _nodeModel.getIndexOf(nodeEntry.getNodeID());
+        if (idx < 0) {
+            return;
+        }
+        NodeEntry last = _nodeModel.getElementAt(idx);
+        if (last != nodeEntry) {
+            // not the same object -- we're talking about an abandoned entry.
+            nodeEntry.dispose();
+            return;
+        }
+        NodeEntry sel = (NodeEntry) _nodeModel.getSelectedItem();
+        _nodeModel.removeElementAt(idx);
+        _nodeModel.insertElementAt(nodeEntry, idx);
+        _nodeModel.setSelectedItem(sel);
+    }
+
+    protected static class NodeEntry implements Comparable<NodeEntry>, PropertyChangeListener {
+        final MimicNodeStore.NodeMemo nodeMemo;
+        String description = "";
+
+        NodeEntry(MimicNodeStore.NodeMemo memo) {
+            this.nodeMemo = memo;
+            memo.addPropertyChangeListener(this);
+            updateDescription();
+        }
+
+        /**
+         * Constructor for prototype display value
+         *
+         * @param description prototype display value
+         */
+        public NodeEntry(String description) {
+            this.nodeMemo = null;
+            this.description = description;
+        }
+
+        public NodeID getNodeID() {
+            return nodeMemo.getNodeID();
+        }
+
+        MimicNodeStore.NodeMemo getNodeMemo() {
+            return nodeMemo;
+        }
+
+        private void updateDescription() {
+            SimpleNodeIdent ident = nodeMemo.getSimpleNodeIdent();
+            StringBuilder sb = new StringBuilder();
+            sb.append(nodeMemo.getNodeID().toString());
+
+            addToDescription(ident.getUserName(), sb);
+            addToDescription(ident.getUserDesc(), sb);
+            if (!ident.getMfgName().isEmpty() || !ident.getModelName().isEmpty()) {
+                addToDescription(ident.getMfgName() + " " +ident.getModelName(), sb);
+            }
+            addToDescription(ident.getSoftwareVersion(), sb);
+            String newDescription = sb.toString();
+            if (!description.equals(newDescription)) {
+                description = newDescription;
+            }
+        }
+
+        private void addToDescription(String s, StringBuilder sb) {
+            if (!s.isEmpty()) {
+                sb.append(" - ");
+                sb.append(s);
+            }
+        }
+
+        private long reorder(long n) {
+            return (n < 0) ? Long.MAX_VALUE - n : Long.MIN_VALUE + n;
```

Comment:
Not sure I understand this.  In the first `Long.MAX_VALUE - n`, `n` is less than zero, i.e. negative. Doesn't subtracting it then overflow?

Reply:
This is from some of the code copied from OpenLCB_Java-master/src/org/openlcb/swing/NodeSelector.java.  

This was cloned in order to filter the node list to only include Tower LCC+Q nodes.